src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+        }
+    }
+
+    fn take(&mut self) -> HANDLE {
+        mem::replace(&mut self.h, INVALID_HANDLE_VALUE)
+    }
+}
+
+// MessageReader implements blocking/nonblocking reads of messages
+// from the handle
+#[derive(Debug)]
+struct MessageReader {
+    // The reader handle; same as the one on OsIpcReceiver, here for
+    // convenience
+    handle: HANDLE,
+
+    // The OVERLAPPED struct for async IO on this receiver; we'll only
+    // ever have one in flight
+    ov: Box<winapi::OVERLAPPED>,
+
+    // A read buffer for any pending reads
+    read_buf: Vec<u8>,
+
+    // If we have already issued an async read
+    read_in_progress: bool,
+
+    // If we received a BROKEN_PIPE or other error
+    // indicating that the remote end has closed the pipe
+    closed: bool,
+
+    // this is ONLY here for debugging purposes.
+    // (for printfs so we know which iocp this is part of)
+    iocp: HANDLE,
+}
+
+impl MessageReader {
+    fn new(handle: HANDLE) -> MessageReader {
+        MessageReader {
+            handle: handle,
+            ov: create_overlapped(),
+            read_buf: Vec::with_capacity(READ_BUFFER_SIZE),
+            read_in_progress: false,
+            closed: false,
+            iocp: 0 as HANDLE,
+        }
+    }
+
+    fn ov_ptr(&self) -> *mut winapi::OVERLAPPED {
+        self.ov.deref() as *const _ as *mut winapi::OVERLAPPED
+    }
+
+    fn reset_overlapped(&mut self) {
+        reset_overlapped(&mut self.ov);
+    }
+
+    fn cancel_io(&mut self) {
+        unsafe {
+            if self.read_in_progress {
+                kernel32::CancelIoEx(self.handle, self.ov_ptr());
+                self.read_in_progress = false;
+            }
+        }
+    }
+
+    fn ensure_is_reading(&mut self) -> Result<(),WinError> {
+        self.start_read()
+    }
+
+    // Called when we receive an IO Completion Packet for this handle.
+    fn notify_completion(&mut self, err: u32) -> Result<(),WinError> {
+        dd2!("[$ {:?}:{:?}] notify_completion", self.iocp, self.handle);
+
+        if err == winapi::ERROR_BROKEN_PIPE {
+            assert!(!self.closed, "we shouldn't get an async BROKEN_PIPE after we already got one");
+            self.closed = true;
+            return Ok(());
+        }
+
+        let nbytes = self.ov.InternalHigh as u32;
+        let offset = self.ov.Offset;
+
+        assert!(offset == 0);
+
+        // if the remote end closed...
+        if err != winapi::ERROR_SUCCESS {
+            // This should never happen
+            panic!("[$ {:?}:{:?}] *** notify_completion: unhandled error reported! {}", self.iocp, self.handle, err);
+        }
+
+        unsafe {
+            let new_size = self.read_buf.len() + nbytes as usize;
+            dd!("nbytes: {}, offset {}, buf len {}->{}, capacity {}",
+                nbytes, offset, self.read_buf.len(), new_size, self.read_buf.capacity());
+            assert!(new_size <= self.read_buf.capacity());
+            self.read_buf.set_len(new_size);
```

Comment:
I don't think you're reading the docs properly -- a vec guarantees that:

> If a Vec has allocated memory, then the memory it points to is on the heap (as defined by the allocator Rust is configured to use by default), and its pointer points to len() initialized elements in order (what you would see if you coerced it to a slice), followed by capacity() - len() logically uninitialized elements.

Also, if you weren't allowed to use the excess capacity, there would be no purpose to the (unsafe) `set_len` function.